File: cpp/650_min_steps/min_steps.cpp
/*
 * @Date: 2021-09-19 09:01:06
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-09-19 09:04:45
 */

#include <cassert>
#include <climits>
#include <vector>

using namespace std;

class Solution {
 public:
  int minSteps(int n) {
    vector<int> f(n + 1);
    for (int i = 2; i <= n; ++i) {
      f[i] = INT_MAX;
      for (int j = 1; j * j <= i; ++j) {
        if (i % j == 0) {
          f[i] = min(f[i], f[j] + i / j);
          f[i] = min(f[i], f[i / j] + j);
        }
      }
    }
    return f[n];
  }
};

int main() {
  assert(Solution().minSteps(3) == 3);
  assert(Solution().minSteps(1) == 0);
}

File: rust/650_min_steps/min_steps.rs
/*
 * @Date: 2021-09-19 09:01:09
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-09-19 09:24:30
 */

struct Solution;

impl Solution {
    pub fn min_steps(n: i32) -> i32 {
        let n = n as usize;
        let mut f: Vec<i32> = vec![0; n + 1];
        for i in 2..=n {
            f[i] = i32::MAX;
            let mut _j = 1;
            while _j * _j <= i {
                if i % _j == 0 {
                    f[i] = std::cmp::min(f[i], f[_j] + i as i32 / _j as i32);
                    f[i] = std::cmp::min(f[i], f[i / _j] + _j as i32);
                }
                _j += 1;
            }
        }
        f[n]
    }
}

fn main() {
    assert_eq!(Solution::min_steps(3), 3);
    assert_eq!(Solution::min_steps(1), 0);
}
